docs/design/fabric-v2+/interfaces.ecc-tlcc-channel.md

```diff
@@ -0,0 +1,534 @@
+ ```
+
+### EDL
+
+#### ECC<->ECC_Enclave
+
+```
+    untrusted {
+        int ocall_tl_session_rpc(
+                [in, size=request_len] uint8_t* request, uint32_t request_len,
+                [out, size=max_response_len] uint8_t* response, uint32_t max_response_len,
+                [out] uint32_t* response_len,
+                [user_check] void *u_shim_ctx);
+    };
+```
+
+#### TLCC<->TLCC_Enclave
+
+```
+    trusted {
+        public int ecall_tl_session_rpc(
+                [in, size=request_len] uint8_t* request, uint32_t request_len,
+                [out, size=max_response_len] uint8_t* response, uint32_t max_response_len,
+                [out] uint32_t* response_len,
+                [user_check] void *u_shim_ctx);
+    };
+```
+
+### Protocol Messages
+
+
+```protobuf
+syntax = "proto3"
+
+message SessionSetupInitRequest {
+        string channel_id    = 1;
+        string chaincode_id  = 2;
+        string enclave_id    = 3;
+}
+
+message SessionSetupInitResponse {
+        uint64 session_id = 1;
+        bytes msg1        = 2; // struct sgx_dh_msg1_t
+}
+
+message SessionSetupCompleteRequest {
+        uint64 session_id = 1;
+        bytes msg2        = 2; // struct sgx_dh_msg2_t
+}
+
+message SessionSetupCompleteResponse {
+        uint64 session_id    = 1;
+        bytes msg3           = 2; // struct sgx_dh_msg2_t
+        string channel_id    = 3;
+        bytes channel_hash   = 4;
+        string chaincode_id  = 5;
+        string enclave_id    = 6;
+}
+
+message SessionCloseRequest {
+        uint64 session_id = 1;
+}
+
+message SessionCloseResponse {
+        uint64 session_id = 1;
+}
+
+message SessionTXRequest {
+        uint64 session_id = 1;
+        bytes nonce       = 2;
+        bytes request     = 3;
+}
+
+message SessionTXResponse {
+        uint64 session_id = 1;
+        bytes nonce       = 2; // should be same as in request: somewhat redundant but makes processing of MAC easier/more uniform.
+        bytes respoonse   = 3;
+}
+
+message SessionError {
+        uint64 session_id = 1;
+        int error_code    = 2;
+        string error_msg  = 3;
+}
+
+message SessionMsg {
+    oneof payload {
+        SessionSetupInitRequest         stpIntReq = 1;
+        SessionSetupInitResponse        stpIntRsp = 2;
+        SessionSetupCompleteRequest     stpCmpReq = 3;
+        SessionSetupCompleteResponse    stpCmpRsp = 4;
+        SessionCloseRequest             clsReq    = 5;
+        SessionCloseResponse            clsRsp    = 6;
+        SessionTXRequest                txReq     = 7;
+        SessionTXResponse               txRsp     = 8;
+        SessionError                    error     = 9;
+    }
+    bytes mac                                    = 99;
+    // MAC is over payload, unless payload is in set
+    //    { SessionSetupInitRequest, SessionSetupInitResponse, SessionSetupCompleteRequest, SessionError }
+    // Note: this replaces the CMAC previously used
```

Comment:
I believe its the MAC over the "serialzed" payload. if we treat the payload as oneof, the receiver of the session msg, will need to get the payload message, serialze it and use it as input to validate the mac. Due to non-deterministic behavior of protobuf serialization there might be an issue with this approach. an alternative would be to introduce yet another sub message and make the payload as bytes in SessionMsg. So the receiver can first check the mac, they parse the payload and continue with processing.

Reply:
it definitely was meant over the serialized payload.  Regarding non-deterministic behaviour. For a given language i think it is deterministic, they just don't guarantee that that different language version guarantee the same as they haven't defined a unique encoding scheme.  So as long as we use the same version we should be fine (which we would do here anyway). It raises though a good question how to best handle it for attestation. Maybe there we should define AttestedData as ANY field in CC_Params so we never have to re-serialize (as-is, you wouldn't really have a serialization of AttestedData only)

Reply:
PS: For uniformity, we could (and maybe should) do payload as ANY (of a separate message containing only the one-of) also here ...

Reply:
hmm, glancing/grepping through the current 0.3 version we mandate, i don't think there is support for ANY, seems we would have to go to 0.4!

Reply:
@mbrandenburger seems moving up to 0.4 to get support for Any and some other improvements (e.g., now a real docu) does work fine with one big caveat (which took me more time to figure out than i would want to admit):  with the latest 0.4.2 version you _must_ use the python 2 generator, not the (default) python3 one or you get some funky python error when it tries to process messages in the two collection.protos....

The diff for 0.4.2 would be as follows (but probably better strategy would be to stay with 0.4.1, report the bug and hope it will get fixed: note that supposedly python3 did work for a long time, it is only in 0.4.2 that the default was switched to explicitly use python3 rather than just use the platform default ...)

```
diff --git a/tlcc_enclave/generate_protos.sh b/tlcc_enclave/generate_protos.sh
index ff20e0d..2445d1a 100755
--- a/tlcc_enclave/generate_protos.sh
+++ b/tlcc_enclave/generate_protos.sh
@@ -32,7 +32,7 @@ fi
 rm -rf $BUILD_DIR
 mkdir -p $BUILD_DIR

-PROTOC_OPTS="--plugin=protoc-gen-nanopb=$NANOPB_PATH/generator/protoc-gen-nanopb"
+PROTOC_OPTS="--plugin=protoc-gen-nanopb=$NANOPB_PATH/generator/protoc-gen-nanopb-py2"

 # compile google protos (timestamp)
 $PROTOC_CMD "$PROTOC_OPTS" --proto_path="protos" --nanopb_out=$BUILD_DIR protos/google/protobuf/*.proto

```